### PR TITLE
ec2_group: cidr_ip validation (#21871).

### LINKED ISSUE #21871: Ec2_group improper error output of CIDR error
##### ISSUE TYPE

 - Feature Idea / Bug 

##### COMPONENT NAME
ec2_group

##### ANSIBLE VERSION

```
ansible --version
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

MAC to AWS 
##### SUMMARY

Some engineers don't validate their CIDRs, *sigh*.

EC2_Group doesn't validate CIDR notations, when given an improper cidr notation from my colleagues you can imagine there are many cidr notations to check. Ansible errors out and doesn't declare that the CIDR is incorrect, but rather says you have hit the AWS limit for SG rules (50 egress/ingress) 
##### STEPS TO REPRODUCE

      - proto: tcp
        from_port: 80
        to_port: 80
        cidr_ip: 10.0.0.28/29

(Proper net cidr is 10.0.0.24/29 but the error doesn't reflect that.

```yaml
---
- name: Some play
  hosts: 127.0.0.1
  connection: local
  gather_facts: false
  tasks:
    - name: <some SG name>
      local_action:
        module: ec2_group
        name: <some SG name>
        profile: <some profile>
        state: present
        vpc_id: <some vpc>
        region: us-east-1
        rules:
      - proto: tcp
        from_port: 80
        to_port: 80
        cidr_ip: 10.0.0.28/29
```

##### EXPECTED RESULTS

To error correct and yell at me for having an improper CIDR notation at X line. 
##### ACTUAL RESULTS

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: <Response><Errors><Error><Code>RulesPerSecurityGroupLimitExceeded</Code><Message>The maximum number of rules per security group has been reached.</Message></Error></Errors><RequestID>c398627e-ab1c-4e58-920c-3492835e02e1</RequestID></Response>
fatal: [127.0.0.1 -> localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/kr/pdz2r4pd43q4f4bwfsx6d_20b2sqwb/T/ansible_6dBl6n/ansible_module_ec2_group.py\", line 483, in <module>\n    main()\n  File \"/var/folders/kr/pdz2r4pd43q4f4bwfsx6d_20b2sqwb/T/ansible_6dBl6n/ansible_module_ec2_group.py\", line 435, in main\n    cidr_ip=thisip)\n  File \"/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/boto-2.43.0-py2.7.egg/boto/ec2/connection.py\", line 3245, in authorize_security_group_egress\n    params, verb='POST')\n  File \"/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/boto-2.43.0-py2.7.egg/boto/connection.py\", line 1227, in get_status\n    raise self.ResponseError(response.status, response.reason, body)\nboto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors><Error><Code>RulesPerSecurityGroupLimitExceeded</Code><Message>The maximum number of rules per security group has been reached.</Message></Error></Errors><RequestID>c398627e-ab1c-4e58-920c-3492835e02e1</RequestID></Response>\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

### PR DESCRIPTION
**Pull request by anryko**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

ec2_group

##### ANSIBLE VERSION

```
ansible 2.3.0 (ec2_group_ip_fix f6805870e9) last updated 2017/02/26 01:55:45 (GMT +200)
```

##### SUMMARY

Fixes #21871.
Added validation of ec2 group cidr_ip before submitting it to boto. Change results in:
 - Cleaner error message on invalid IP.
 - Automatically adjusts CIDR IP in case when interface IP was provided (e.g. 10.0.0.28/29 will be adjusted to 10.0.0.24/29). Before the change boto was doing the same thing internally and rule was set, but would fail on second run.
 - /32 is automatically appended to IP if netmask is not set.

##### DETAILS
Rule:
```
      - proto: tcp
        from_port: 22
        to_port: 22
        cidr_ip: 10.0.0.28/29
```

Before the change:
First run:
```
changed: [localhost] => {                                                                                             [60/216]
    "changed": true,
    "group_id": "sg-xxxxxxxx",
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "description": "edge-test",
            "ec2_url": null,
            "name": "edge-test",
            "profile": null,
            "purge_rules": true,
            "purge_rules_egress": true,
            "region": "eu-central-1",
            "rules": [
                {
                    "cidr_ip": "10.0.0.28/29",
                    "from_port": 22,
                    "proto": "tcp",
                    "to_port": 22
                }
            ],
            "rules_egress": null,
            "security_token": null,
            "state": "present",
            "validate_certs": true,
            "vpc_id": "vpc-xxxxxxxx"
        },
        "module_name": "ec2_group"
    }
}
```
Secind run:
```
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_ayE3Gh/ansible_module_ec2_group.py", line 483, in <module>
    main()
  File "/tmp/ansible_ayE3Gh/ansible_module_ec2_group.py", line 377, in main
    group.authorize(rule['proto'], rule['from_port'], rule['to_port'], thisip, grantGroup)
  File "/usr/local/lib/python2.7/dist-packages/boto/ec2/securitygroup.py", line 203, in authorize
    dry_run=dry_run)
  File "/usr/local/lib/python2.7/dist-packages/boto/ec2/connection.py", line 3207, in authorize_security_group
    params, verb='POST')
  File "/usr/local/lib/python2.7/dist-packages/boto/connection.py", line 1227, in get_status
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidPermission.Duplicate</Code><Message>the specified rule "peer: 10.0.0.24/29, TCP, from p$
rt: 22, to port: 22, ALLOW" already exists</Message></Error></Errors><RequestID>8ad2ecaa-a82c-4618-910f-277b0f15c009</Request$
D></Response>

fatal: [localhost]: FAILED! => {
    "changed": false,
    "failed": true,
    "invocation": {
        "module_name": "ec2_group"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_ayE3Gh/ansible_module_ec2_group.py\", line 48$
, in <module>\n    main()\n  File \"/tmp/ansible_ayE3Gh/ansible_module_ec2_group.py\", line 377, in main\n    group.authorize$
rule['proto'], rule['from_port'], rule['to_port'], thisip, grantGroup)\n  File \"/usr/local/lib/python2.7/dist-packages/boto/$
c2/securitygroup.py\", line 203, in authorize\n    dry_run=dry_run)\n  File \"/usr/local/lib/python2.7/dist-packages/boto/ec2$
connection.py\", line 3207, in authorize_security_group\n    params, verb='POST')\n  File \"/usr/local/lib/python2.7/dist-pack
ages/boto/connection.py\", line 1227, in get_status\n    raise self.ResponseError(response.status, response.reason, body)\nbot
o.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors
><Error><Code>InvalidPermission.Duplicate</Code><Message>the specified rule \"peer: 10.0.0.24/29, TCP, from port: 22, to port:
 22, ALLOW\" already exists</Message></Error></Errors><RequestID>8ad2ecaa-a82c-4618-910f-277b0f15c009</RequestID></Response>\n
",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
```

After the change:
First run:
```
changed: [localhost] => {                                                                                             [4/9395]
    "changed": true,
    "group_id": "sg-xxxxxxxx",
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "description": "edge-test",
            "ec2_url": null,
            "name": "edge-test",
            "profile": null,
            "purge_rules": true,
            "purge_rules_egress": true,
            "region": "eu-central-1",
            "rules": [
                {
                    "cidr_ip": "10.0.0.24/29",
                    "from_port": 22,
                    "proto": "tcp",
                    "to_port": 22
                }
            ],
            "rules_egress": null,
            "security_token": null,
            "state": "present",
            "validate_certs": true,
            "vpc_id": "vpc-xxxxxxxx"
        },
        "module_name": "ec2_group"
    }
}
```
Second run:
```
ok: [localhost] => {                                                                                                  [4/9446]
    "changed": false,
    "group_id": "sg-xxxxxxxx",
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "description": "edge-test",
            "ec2_url": null,
            "name": "edge-test",
            "profile": null,
            "purge_rules": true,
            "purge_rules_egress": true,
            "region": "eu-central-1",
            "rules": [
                {
                    "cidr_ip": "10.0.0.24/29",
                    "from_port": 22,
                    "proto": "tcp",
                    "to_port": 22
                }
            ],
            "rules_egress": null,
            "security_token": null,
            "state": "present",
            "validate_certs": true,
            "vpc_id": "vpc-xxxxxxxx"
        },
        "module_name": "ec2_group"
    }
}
```
Notice that before the change "incorrect" IP is in the output and we can see the IP that was actually set in error message on the second run.

### CODE REVIEW

**Comment by willthames** [comment]
I wonder if a `cidr` parameter type would be a better approach that could be more widely applied, similar to how the `path` type expands `~` in filenames (see https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1659 for how that happens).

**Comment by anryko** [comment]
@willthames that's a good idea for a new feature. However, it is not applicable to this particular case. As far as I encountered in ansible sources, validation is done only on top level, like:
```python
argument_spec.update(dict(
   rules=dict(type='list')
))
```
In this case we would need "deep" validation of list of dicts of rules, which can contain cidr_ip (or not). This might be another nice feature to have, but I think it is out of scope of this small fix.

**Comment by sivel** [comment]
Some of that discussion is ongoing, such as in:

https://github.com/ansible/ansible/pull/17731
https://github.com/ansible/ansible/pull/21171

However the implementation in #17731 isn't what we want.

For now, I'd probably just leave it implemented the way it is here, and look for this in the future.

My biggest concern is that this is an additional dependency that could break backwards compatibility.

You should like wrap the import and define a `HAS_IPADDRESS` and if `HAS_IPADDRESS` is false, don't validate it, but instead maybe add a warning to the return, saying it couldn't be validated due to missing the `ipaddress` python module.

Also, I might recommend updating to utilize `netaddr` since that is already the library that the `ipaddr` filters are based off of, and what is more likely to be available.

**Comment by anryko** [comment]
@sivel thanks for your review and links. Will come back with recommended changes.

File: lib/ansible/modules/cloud/amazon/ec2_group.py [hunk 1]
```diff
@@ -133,6 +133,8 @@
         group_desc: other example EC2 group
 '''
 
+import ipaddress
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_group.py]: I believe this module is only available in 2.7 and 3.3+, for modules using boto we support back to 2.6. How is boto doing this validation on Python 2.6? Can you copy them?

**Review by ryansb — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_group.py (hunk 1 above)
Comment (anryko) [reply lib/ansible/modules/cloud/amazon/ec2_group.py]: @ryansb checked it out and boto doesn't do cidr_ip validation. So this validation is done on AWS backend.

**Comment by anryko** [comment]
@sivel will use any of those libs if available and display warning while skipping validation if not. What do you think?

**Comment by s-hertel** [comment]
Thanks for working on this @anryko. Rather than depend on external libraries, this was added in d877c146ab750c0f1505b4180e4170163d854fb5 using module_utils/network/common/utils.py.

**Pull request closed without merging**